### DISCUSSION TITLE
Adding custom attributes to user consent.

### DISCUSSION THREAD

**Discussion by ramonpenteado** [keycloak/keycloak#23139] [Q&A] [answered]
Recently I posted a question here about how to add a custom claim to access_token. The answer was pretty helpful and helped us to get where we are now. After being able to retrieve this attribute (in this case it is `authorization_details` from a `PAR` request) and adding it to the access_token, we would like to somehow retrieve this information for a custom consent page.
So, basically, thirty-party-app send `authorization_details` at the body -> Keycloak creates a access_token with this `authorization_details` claim -> custom consent page (to which user gets redirect) retrieves this `authorization_details` somewhere.
Is there a endpoint, or a SPI we could use to achieve that?

Reference of the issue I was answered: https://github.com/keycloak/keycloak/discussions/22991

**Comment by justin-tay** [comment]
At this point I'm not sure if what you are trying to do is a good idea, because essentially what you want seems to be support in Keycloak for `authorization_details`. You can attempt to hack it in, but you'd be relying on existing implementation details in Keycloak, and certain things may or may not be possible, like I'm not sure if you can save the consent, and in fact I'm not sure at the end of the day this is going to meet your requirements. A pull request to Keycloak may be required to meet your needs.

The flow is more like
- Application sends `authorization_details` in PAR request
- Consent form generated for user to accept when user redirected to authorization endpoint
- Access token generated at token endpoint when application presents code

To customize the inputs to the consent form you could have a custom `org.keycloak.forms.login.LoginFormsProviderFactory` that extends off `org.keycloak.forms.login.freemarker.FreeMarkerLoginFormsProviderFactory`. The code would resemble something similar to the below.

```java
public class CustomLoginFormsProvider extends FreeMarkerLoginFormsProvider {
	public CustomLoginFormsProvider(KeycloakSession session) {
		super(session);
	}

	@Override
	protected Response createResponse(LoginFormsPages page) {
		if(LoginFormsPages.OAUTH_GRANT.equals(page)) {
			String authorizationDetails = this.authenticationSession.getClientNotes()
					.get("client_request_param_authorization_details");
                        // Set the json from the authorizationDetails
			AuthorizationDetailsJSONRepresentation json = new AuthorizationDetailsJSONRepresentation();
			json.setCustomData("action", xxx);
			json.setCustomData("scope_parameter", xxx);
			json.setType(AuthorizationDetailsJSONRepresentation.DYNAMIC_SCOPE_RAR_TYPE);
			ClientScopeModel clientScopeModel = new MyClientScopeModel(json);
			AuthorizationDetails details = new AuthorizationDetails(clientScopeModel, AuthorizationRequestSource.SCOPE, json);
			this.clientScopesRequested.add(details);
		}
		return super.createResponse(page);
	}
}
```

The related freemarker template is `login-oauth2-grant.ftl`.

https://github.com/keycloak/keycloak/blob/56b94148a0e05c89fc79e2eecc6790565db05fb1/themes/src/main/resources/theme/base/login/login-oauth-grant.ftl#L16-L30

You also probably need to pay attention to these sets of codes

https://github.com/keycloak/keycloak/blob/56b94148a0e05c89fc79e2eecc6790565db05fb1/services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java#L1116-L1130

**Reply by ramonpenteado** [reply]
So, you mean that it is not a good idea to have `authorization_details` be sent to a custom consent page?
Because Im not entirely interested in having the Consent save. I preferred to some how retrieve the `authorization_details` at the time my user hits Custom Consent Page. We could handle this "save" else where possibly, other than inside Keycloak.
Did I get it right?

**Reply by justin-tay** [reply, marked as answer]
What I mean is that I think that when implementing this you may need to rely on certain implementation details that you cannot customize as part of an SPI. For instance you may be relying on portions of logic designed for the dynamic scope implementation. You are essentially developing around what is currently implemented. As I said you could implement a custom `org.keycloak.forms.login.LoginFormsProviderFactory` to set the attributes you require on the consent page but the implementation will look awkward.

You may want to do a status check on the following before proceeding
- https://github.com/tnorimat/keycloak/pull/24
- https://github.com/keycloak/keycloak-community/pull/266

**Reply by ramonpenteado** [reply]
Thanks Justin
Yes, those PRs seems to be very close to solve this issue for me.
We'll keep an eye on it. Thank you very much for your help.
